File: engine/src/main/java/com/s3bender/model/BucketEntity.java
package com.s3bender.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;

import java.time.Instant;

@Entity
@Table(name = "buckets", uniqueConstraints = @UniqueConstraint(columnNames = "accessKey"))
public class BucketEntity {

    /** Bucket name is the primary key; also doubles as the URL path segment. */
    @Id
    @Column(length = 63)
    private String name;

    @Column(nullable = false, length = 64)
    private String accessKey;

    /** Base64 AES-GCM ciphertext of the bucket's secret key, encrypted with the server master key. */
    @Column(nullable = false, length = 512)
    private String encryptedSecretKey;

    @Column(nullable = false)
    private Instant createdAt;

    protected BucketEntity() {
        // JPA
    }

    public BucketEntity(String name, String accessKey, String encryptedSecretKey, Instant createdAt) {
        this.name = name;
        this.accessKey = accessKey;
        this.encryptedSecretKey = encryptedSecretKey;
        this.createdAt = createdAt;
    }

    public String getName() {
        return name;
    }

    public String getAccessKey() {
        return accessKey;
    }

    public String getEncryptedSecretKey() {
        return encryptedSecretKey;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }
}

File: engine/Models/BucketEntity.cs
using System.ComponentModel.DataAnnotations;

namespace S3Bender.Api.Models;

public class BucketEntity
{
    [Key, MaxLength(63)]
    public string Name { get; set; } = default!;

    /// <summary>Optional free-text note set at creation time; purely for humans browsing the console.</summary>
    [MaxLength(200)]
    public string? Description { get; set; }

    [Required, MaxLength(64)]
    public string AccessKey { get; set; } = default!;

    /// <summary>Base64 AES-GCM ciphertext of the bucket's secret key, encrypted with the server master key.</summary>
    [Required, MaxLength(512)]
    public string EncryptedSecretKey { get; set; } = default!;

    public DateTimeOffset CreatedAt { get; set; }
}
